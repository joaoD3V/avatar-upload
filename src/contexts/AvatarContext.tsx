import { createContext, useState, useCallback } from 'react';

export type AvatarStatus = 'initial' | 'editing' | 'saved' | 'error';

export type ImageFileProps = string | ArrayBuffer | undefined;

type AvatarContexType = {
  status: AvatarStatus;
  avatarImage: ImageFileProps;
  sliderValue: number;
  handleClose: () => void;
  handleChangeStatus: (newStatus: AvatarStatus) => void;
  handlePreviewProfileImage: (files: File[]) => void;
  handleUpdateSliderValue: (value: number | number[]) => void;
};

type AvatarContextProviderProps = {
  children: React.ReactNode;
};

export const AvatarContext = createContext({} as AvatarContexType);

export function AvatarContextProvider({
  children,
}: AvatarContextProviderProps) {
  const [status, setStatus] = useState<AvatarStatus>('initial');
  const [avatarImage, setAvatarImage] = useState<ImageFileProps>(undefined);
  const [sliderValue, setSliderValue] = useState<number>(0);

  const handleClose = useCallback(() => {
    setStatus('initial');
    setAvatarImage(undefined);
  }, []);

  const handleChangeStatus = useCallback((newStatus: AvatarStatus) => {
    setStatus(newStatus);
  }, []);

  const handlePreviewProfileImage = useCallback((files: File[]) => {
    if (files && files[0]) {
      const reader = new FileReader();
      reader.addEventListener('load', (event) => {
        const image = event.target!.result!;

        const imageCopy = new Image();
        imageCopy.src = `${image}`;

        imageCopy.onload = () => {
          setAvatarImage(image);
          setSliderValue(0);
          return setStatus('editing');
        };
      });

      reader.readAsDataURL(files[0]);
    }
  }, []);

  const handleUpdateSliderValue = useCallback((value: number | number[]) => {
    setSliderValue(Number(value));
  }, []);

  return (
    <AvatarContext.Provider
      value={{
        status,
        avatarImage,
        sliderValue,
        handleClose,
        handleChangeStatus,
        handlePreviewProfileImage,
        handleUpdateSliderValue,
      }}
    >
      {children}
    </AvatarContext.Provider>
  );
}
