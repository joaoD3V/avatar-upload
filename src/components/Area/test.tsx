import { screen } from '@testing-library/react';
import { AvatarContext, AvatarContextProvider } from 'contexts/AvatarContext';
import { renderWithTheme } from 'utils/tests/helpers';

import { Area } from '.';
import React from 'react';

describe('<Area />', () => {
  it('should render with border and without close button by default', () => {
    const { container } = renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('initial')}
              <Area>Test</Area>
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.queryByLabelText(/close button/i)).not.toBeInTheDocument();
    expect(screen.getByLabelText(/wrapper/i)).toHaveStyle({
      border: '2px dashed #C7CDD3',
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render with close button when it passed', () => {
    renderWithTheme(<Area hasClose>Teste</Area>);

    expect(screen.getByLabelText(/close button/i)).toBeInTheDocument();
  });

  it('should render without border when status is editing', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('editing')}
              <Area>Test</Area>
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.getByLabelText(/wrapper/i)).toHaveStyle({
      border: '0',
    });
  });

  it('should render without border when status is error', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('error')}
              <Area>Test</Area>
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.getByLabelText(/wrapper/i)).toHaveStyle({
      border: '0',
    });
  });

  it('should render with border when status is saved', () => {
    renderWithTheme(
      <AvatarContextProvider>
        <AvatarContext.Consumer>
          {(value) => (
            <>
              {value.handleChangeStatus('saved')}
              <Area>Test</Area>
            </>
          )}
        </AvatarContext.Consumer>
      </AvatarContextProvider>
    );

    expect(screen.getByLabelText(/wrapper/i)).toHaveStyle({
      border: '2px dashed #C7CDD3',
    });
  });
});
