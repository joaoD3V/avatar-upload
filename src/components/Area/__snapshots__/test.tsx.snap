// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Area /> should render with border and without close button by default 1`] = `
.c0 {
  width: 553px;
  height: 177px;
  background: #F2F5F8;
  border: 2px dashed #C7CDD3;
  border-radius: 8px;
  padding: 32px;
  position: relative;
}

@media (max-width:768px) {
  .c0 {
    width: 90vw;
    height: 277px;
    padding: 16px;
  }
}

<div
  aria-label="Wrapper"
  class="c0"
>
  Test
</div>
`;
